### ISSUE TITLE
Recent update to Shiny 0.9.0  broke DataTable input

### ISSUE THREAD

**Issue by pieterjanvc** [posit-dev/py-shiny#1345]
Hello,

The recent update to Shiny 0.9.0 broke the DataTable input, which was one of the updates in this new version.
At this point `input.<tableID>_selected_rows()` stopped working and is no longer updating any reactive function. 
You can check [your own demo example](https://shinylive.io/py/editor/#code=NobwRAdghgtgpmAXAAjFADugdOgnmAGlQGMB7CAFzkqVQDMAnUmZdKAG3gfWoHMBXAJYQAzskEx0pBhWTtSUACYB9HhAHCRAHQiNmyEQAthucZOmyG1RXAY69LIyaxwAHuisixEqTPER0fgoiIR0dNQ1RZABeOQUVCKFRAAoASjCIISxDACZkrTAABQ4uZEK+JO0wdIgMgAErCBsGLFCmuDpkJgB3ETTEHWQhrtJemP9AihwKzWVFOmUROHY4YioVHr7UoYBiZAAefcHhpZW1uEVxgqICrAArUmFkkQoGZMFtumlxfwMLC+Sm1S20EnU2yGWS2QBQAcuQ4AVjkMrBR+AwIMg6AUAEqjMSnVbrFAgAnnRQAX0RtWpDWstiwiigFCgykYsAR7U6iVm836SK6cFR6IF7RaABEmVAACpQABGK2S3NERB6i2WhME5GUMFINmiBREwl4KwK22Qe0OOkIqAouB4tDAVFcFDA5IAukA) provided in the [documentation](https://shiny.posit.co/py/components/outputs/datatable/index.html#select-rows) to see it is broken.

I hope this can be patched soon :) Downgrading to version 0.8.1 reverts the issue.

Thanks!

**Comment by gadenbuie** [comment]
Shiny 0.9.0 introduced many new DataTable features, which includes new syntax for retrieving the selected rows. The new syntax is much more flexible in the sense that it both allows finer grained control over what types of selections are allowed and makes it possible to retrieve the selected rows, columns or cells on the server side.

The best example currently lives in the API reference: https://shiny.posit.co/py/api/core/render.DataGrid.html. We definitely need to update the examples in the components gallery to match the new syntax.

Here's [an updated version of the demo example](https://shinylive.io/py/editor/#code=NobwRAdghgtgpmAXGKAHVA6VBPMAaMAYwHsIAXOcpMAMwCdiYACVKAG3jtUoHMBXAJYQAzkwExUxOmSZtiUACYB9bhH5DhAHQj1GTYQAsh2MRKky6lBXDrbdzQ8YxwAHqkvDR4ydLERUfGR4TILa2qrqIkwAvLLyyhGCIgAUAJRhEIIYBgBMyZpgAArsnEyFvElaYOkQGQAClhDWdBihTXA0TAwA7sJpiNpMQ13EvTEsFRpKCjQYClBkUEoAbgJw3cnCcGxwhBQK0QAqdHxwqUMAxEwAPNeDw1s7e3AK4wXBBRgAVsRCm2R0ZICc40KRiPz6cwvZI9YQYITWFypc4CToQYgWUZwuASMgmbZbJgFABypDgBXuQ0sZD4dAgTBoBQASlj9NtdvtEEwQI8OS8AL4U2rChpWGxzBZLeiwcntTqJKYzfqUrpwGl01XtFoAEUlhygACMdskFSJgry9gJSEoYMRrNECrCCucmFdbtowPyALpAA) you linked to:

```python
from palmerpenguins import load_penguins
from shiny import render
from shiny.express import input, ui

penguins = load_penguins()

ui.h2("Palmer Penguins")


@render.ui
def rows():
    rows = penguins_df.data_view(selected=True)  # <<
    selected = ", ".join(str(i) for i in sorted(rows.index)) if not rows.empty else "None"
    return f"Rows selected: {selected}"


@render.data_frame
def penguins_df():
    return render.DataTable(penguins, selection_mode="rows")  # <<
```

Notice that:

1. In `render.DataTable()` we now use `selection_mode` to specify what selections types are allowed.
2. Rather than using `input.{data_frame_id}_selected_rows()`, we now let you refer to the output directly, with a much richer API for interacting with the data table. Above, I used `penguins_df.data_view(selected=True)` to get the selected rows from the current view (which may be filtered).
3. The return value from the `.data_view()` method is a complete table, which is generally more useful than the indices alone. That said, you can still use `rows.index` to get the indices, which were used in the example to just list which rows were selected.

**Comment by pstorozenko** [comment]
I'd like to throw my bit here.

1. I definitely prefer `.data_view(...)` accessor than some _random_ `*_selected_rows()` :D Thanks for working on this!
2. However, with `_selected_rows()` it was possible to get the indices of selected rows easily.  Now I have to rely on pandas index for this. Why it's important to have such feature? Sometimes the displayed dataframe shows just a subset of columns of some dataframe in memory. Or rows in the dataframe correspond to some other objects (in my particular case each row represents an image, but you get the abstraction). It's not a problem for now, but if one day we shiny switches to some dataframe representation without creature like `pd.Index` it may be.
3. ☝🏻 The breaking changes section of 0.9.0 release mentions only `selection_mode` (which is actually deprecation, not breaking), but says nothing about `_selected_rows()` / `.data_view()` which made it much harder for me to trace the issue.  

Cheers!

**Comment by sidviny** [comment]
Hi,

None of the examples with the DataGrid, nor when I am creating a dashboard example with "shiny create" show a datagrid.

My default environment was basically:
- python 3.12
- pandas 2.2
- shiny: 0.9.0

Reverting back to shiny 0.8.1 did not work

I then created a spin-off of my big production conda environment which has:
- python 3.11
- pandas 2.2

I added shiny, but still, none of these environments show the data grid.

I am a bit clueless why, but, we often have problems with downloading as we have self-signed certificates in some of our firewalls.  We have to add the SSL exceptions for instance when we use pip.  I think this could be related.

**Comment by schloerke** [comment]
@pieterjanvc 
> At this point `input.<tableID>_selected_rows()` stopped working and is no longer updating any reactive function.

Yes. It was removed without warning. Sorry 'bout that! 

I'll restore it for the next release and add a clear news item about removing it in a future release.

> [I'd like to have access to the selected rows]

Good news...

With a combination of 
https://github.com/posit-dev/py-shiny/blob/3b742a2240774521a16c688e9d73c3a80206ce43/shiny/render/_data_frame.py#L405-L412 and https://github.com/posit-dev/py-shiny/blob/3b742a2240774521a16c688e9d73c3a80206ce43/shiny/render/_data_frame.py#L492-L502, maybe we expose them via `.data_view_indices(*,  selected: bool = False)`? (Similar to `.data_view(*, selected: bool = False)`)

It could return a `{rows: list[int], cols: list[int]}` object containing all rows/cols being used (given `selected` value). We could add a `type` to describe how the values were selected, similar to the `CellSelection` type https://github.com/posit-dev/py-shiny/blob/3b742a2240774521a16c688e9d73c3a80206ce43/shiny/render/_data_frame_utils/_selection.py#L162-L185 . 

For the return value, I still don't know if `rows` and `cols` should always exist (containing all values). If they always exist, it is easier to code as a user. If they do not exist, it is much clearer as to what is being subsetted but more annoying to integrate with as a user.  I'm currently leaning towards always including `rows` and `cols` containing all values being used. Ex: multi row selection would return `{type: "row", rows: [1,3,5], cols: [0,1,2,3,4,5]}` assuming we selected rows 1, 3, and 5 and there were 6 columns.

```python
class DataViewIndices(TypedDict):
    type: Literal["row", "col", "rect", "none"]
    rows: ListOrTuple[int]
    cols: ListOrTuple[int]

# class data_frame...
def data_view_indices(self, *, selected: bool = False) -> DataViewIndices:
    ...
```

-----------------------

@pstorozenko Glad you like `.data_view()`!

> It's not a problem for now, but if one day we shiny switches to some dataframe representation without creature like `pd.Index` it may be.

Can I get a copy of your crystal ball? 😆 

I'd love to move to something like https://github.com/data-apis/dataframe-api-compat , however to [get typing support](https://github.com/data-apis/dataframe-api-compat/blob/main/contributing.md#type-checking), their typing specification is defined deep within another repo: https://github.com/data-apis/dataframe-api/tree/main/spec/API_specification/dataframe_api .  

I do not understand why the types being used within `dataframe-api-compat` are not shipped with the package. </rant>

**Comment by schloerke** [comment]
Original issue is being fixed in #1377 . 

Making new issue for `.data_view_info()` support
